cstrml/storage_order/src/lib.rs

```diff
@@ -0,0 +1,134 @@
+        // transfer the fee and return order id
+        0
+    }
+}
+
+/// The module's configuration trait.
+pub trait Trait: system::Trait + tee::Trait {
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+    type OnOrderStroage: OnOrderStroage<Self::AccountId>;
+}
+
+// TODO: add add_extra_genesis to unify chain_spec
+// This module's storage items.
+decl_storage! {
+    trait StoreOrder for Module<T: Trait> as StoreOrder {
+        pub StorageOrders get(fn storage_orders) config(): map (T::AccountId, u64) => Option<StorageOrder<T::AccountId>>; // Cannot use MerkleRoot as the key. cannot open the apps
+    }
+}
+
+// The module's dispatchable functions.
+decl_module! {
+    /// The module declaration.
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // Initializing events
+        // this is needed only if you are using events in your module
+        fn deposit_event() = default;
+
+        fn store_storage_order(
+            origin,
+            dest: <T::Lookup as StaticLookup>::Source,
+            #[compact] value: Balance, // Keep it now and refactor it later
+            file_indetifier: MerkleRoot,
+            file_size: u64,
+            expired_duration: BlockNumber,
+            expired_on: BlockNumber
+        ) -> DispatchResult
+            {
+                let who = ensure_signed(origin)?;
+                let dest = T::Lookup::lookup(dest)?;
+                let storage_order = StorageOrder::<T::AccountId> {
+                    file_indetifier,
+                    file_size,
+                    expired_duration,
+                    expired_on,
+                    destination: dest.clone()
+                };
+                // 1. Do check and should do something
+
+                ensure!(Self::storage_order_check(&storage_order).is_ok(), "Storage Order is invalid!");
+
+                Self::transfer_fee_and_store_order(&who, &dest, value, &storage_order);
+                
+                // 3. Emit storage order event
+                Self::deposit_event(RawEvent::ReportStorageOrders(who, storage_order));
+
+                Ok(())
+            }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    // private function can be built in here
+    fn storage_order_check(so: &StorageOrder<T::AccountId>) -> DispatchResult {
+        ensure!(
+            &<tee::Module<T>>::get_last_work_report(&so.destination).is_some(),
+            "Cannot find work report!"
+        );
+        ensure!(
+            &<tee::Module<T>>::get_last_work_report(&so.destination).unwrap().empty_workload >= &so.file_size,
+            "Empty work load is not enough!"
+        );
+        Ok(())
+    }
+
+    fn transfer_fee_and_store_order(source: &T::AccountId, dest: &T::AccountId, value: Balance, so: &StorageOrder<T::AccountId>) {
+        let fee_id = T::OnOrderStroage::storage_fee_transfer(&source, &dest, value);
```

Comment:
maybe considering just rename to `maybe_insert_sorder`? we can use `sorder` and `rorder` to simplify `storage order` and `retrieval order`, this maybe shorten func names and variable names